### PR TITLE
fix(rule generation): redirect to melange.emit parents 

### PR DESCRIPTION
**Pull request by anmonteiro**
This is just the rule generation / redirection part of https://github.com/ocaml/dune/pull/6541 without the removal of the `(target ..)` field.

The test added in #6506 is now fixed with this change.

### CODE REVIEW

File: src/dune_rules/gen_rules.ml [hunk 1]
```diff
@@ -374,6 +374,35 @@ let has_rules subdirs f =
...
+        match
+          List.exists stanzas ~f:(function
+            | Melange_emit _ -> true
+            | _ -> false)
+        with
+        | true -> Memo.return true
+        | false -> does_parent_have_melange_emit rest))
+  in
+
+  does_parent_have_melange_emit (List.rev (all_parents dir))
```
Comment (rgrinberg) [inline src/dune_rules/gen_rules.ml:404]: Simpler would be to pass the directory directly and call `Path.Build.parent` until it returns `None`.

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: That is, in fact, a lot better. done in [e9f9966](https://github.com/ocaml/dune/pull/6579/commits/e9f9966255f3677a61cd526cef0d43ca31889ae6)

File: src/dune_rules/gen_rules.ml [hunk 2]
```diff
@@ -374,6 +374,21 @@ let has_rules subdirs f =
...
+let rec is_under_melange_emit dir =
+  match Path.Build.parent dir with
+  | None -> Memo.return false
+  | Some dir -> (
+    Only_packages.stanzas_in_dir dir >>= function
+    | None -> is_under_melange_emit dir
+    | Some { Dune_file.stanzas; dir = _; project = _ } -> (
+      match
+        List.exists stanzas ~f:(function
+          | Melange_emit _ -> true
```
Comment (rgrinberg) [inline src/dune_rules/gen_rules.ml:386]: This check isn't quite right. We should be checking if we're under `$directory-of-melanage-stanzar/$target-field-of-melange-stanza`

Comment (anmonteiro) [reply src/dune_rules/gen_rules.ml]: Good point. I think the current check is only valid after we remove the `target` field in https://github.com/ocaml/dune/pull/6541.

Pushed [5216189](https://github.com/ocaml/dune/pull/6579/commits/5216189d8dca993c5101de3116fb03550480e231)

**Comment by anmonteiro** [comment]
Previously broken tests are now fixed. Seeing the same (puzzling to me) failures as the main branch

**Comment by rgrinberg** [comment]
> Seeing the same (puzzling to me) failures as the main branch

Are you talking about the CI errors?

**Comment by anmonteiro** [comment]
Yes, some errors in the diffing

**Comment by rgrinberg** [comment]
They're not related to this PR. They are appearing in other PR's as well

Rudi.

**Comment by anmonteiro** [comment]
I noticed that. I mentioned they’re present in `main` in my original message

**Pull request merged**